MongoDB Connection State is not 'open'

## How  the Problem was Produced

The source mongodb has 500K matching trip records.
The following returns one of these records without issue.
```
CALL apoc.mongodb.first('mongodb://localhost:27017', 'mobilytics', 'mds_trips', {provider_name: 'Lime'}, false) YIELD value as trip;
```
The following request produces the error.

```
CALL apoc.mongodb.find('mongodb://localhost:27017', 'mobilytics', 'mds_trips', {provider_name: 'Lime'}, null, null, false, 0, 0) YIELD value as trip ;
```

Failed to invoke procedure `apoc.mongodb.find`: Caused by: java.lang.IllegalStateException: state should be: open

### Versions

- OS: Ubuntu 18.04
- Neo4j: 3.5.3
- Neo4j-Apoc: apoc-3.5.0.3-all.jar 

- bson-3.5.0.jar  
-  mongo-java-driver-3.5.0.jar
-  mongodb-driver-3.5.0.jar
-  mongodb-driver-core-3.5.0.jar

- jackson-annotations-2.9.8.jar
-  jackson-core-2.9.8.jar
-  jackson-databind-2.9.8.jar

--- a/src/main/java/apoc/mongodb/MongoDB.java
+++ b/src/main/java/apoc/mongodb/MongoDB.java
@@ -14,6 +14,8 @@ import java.io.Closeable;
 import java.lang.reflect.InvocationTargetException;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
 import java.util.stream.Stream;
 
 /*
@@ -57,24 +59,20 @@ public class MongoDB {
                                  @Name("query") Map<String, Object> query,
                                  @Name(value = "compatibleValues", defaultValue = "false") boolean compatibleValues, @Name(value = "skip", defaultValue = "0") Long skip,
                                  @Name(value = "limit", defaultValue = "0") Long limit) {
-        try (Coll coll = getMongoColl(hostOrKey, db, collection, compatibleValues)) {
-            return coll.all(query, skip, limit).map(MapResult::new); // .onClose(coll::safeClose);
-        } catch (Exception e) {
-            log.error("apoc.mongodb.get - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], compatibleValues = [" + compatibleValues + "], skip = [" + skip + "], limit = [" + limit + "]", e);
-            throw new RuntimeException(e);
-        }
+        return executeMongoQuery(hostOrKey, db, collection, compatibleValues,
+                coll -> coll.all(query, skip, limit).map(MapResult::new),
+                e -> log.error("apoc.mongodb.get - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], compatibleValues = [" + compatibleValues + "], skip = [" + skip + "], limit = [" + limit + "]", e));
     }
 
     @Procedure
     @Description("apoc.mongodb.count(host-or-port,db-or-null,collection-or-null,query-or-null) yield value - perform a find operation on mongodb collection")
     public Stream<LongResult> count(@Name("host") String hostOrKey, @Name("db") String db, @Name("collection") String collection, @Name("query") Map<String, Object> query) {
-        try (Coll coll = getMongoColl(hostOrKey, db, collection, false)) {
-            long count = coll.count(query);
-            return Stream.of(new LongResult(count)); // .onClose(coll::safeClose);
-        } catch (Exception e) {
-            log.error("apoc.mongodb.count - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "]",e);
-            throw new RuntimeException(e);
-        }
+        return executeMongoQuery(hostOrKey, db, collection, false,
+                coll -> {
+                    long count = coll.count(query);
+                    return Stream.of(new LongResult(count));
+                },
+                e -> log.error("apoc.mongodb.count - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "]",e));
     }
 
     private Coll getColl(@Name("host") String hostOrKey, @Name("db") String db, @Name("collection") String collection, boolean compatibleValues) {
@@ -85,13 +83,12 @@ public class MongoDB {
     @Procedure
     @Description("apoc.mongodb.first(host-or-port,db-or-null,collection-or-null,query-or-null,[compatibleValues=true|false]) yield value - perform a first operation on mongodb collection")
     public Stream<MapResult> first(@Name("host") String hostOrKey, @Name("db") String db, @Name("collection") String collection, @Name("query") Map<String, Object> query, @Name(value = "compatibleValues", defaultValue = "false") boolean compatibleValues) {
-        try (Coll coll = getMongoColl(hostOrKey, db, collection, compatibleValues)) {
-            Map<String, Object> result = coll.first(query);
-            return result == null || result.isEmpty() ? Stream.empty() : Stream.of(new MapResult(result)); // .onClose(coll::safeClose);
-        } catch (Exception e) {
-            log.error("apoc.mongodb.first - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], compatibleValues = [" + compatibleValues + "]",e);
-            throw new RuntimeException(e);
-        }
+        return executeMongoQuery(hostOrKey, db, collection, compatibleValues,
+                coll -> {
+                    Map<String, Object> result = coll.first(query);
+                    return result == null || result.isEmpty() ? Stream.empty() : Stream.of(new MapResult(result));
+                },
+                e -> log.error("apoc.mongodb.first - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], compatibleValues = [" + compatibleValues + "]",e));
     }
 
     @Procedure
@@ -105,12 +102,9 @@ public class MongoDB {
                                   @Name(value = "compatibleValues", defaultValue = "false") boolean compatibleValues,
                                   @Name(value = "skip", defaultValue = "0") Long skip,
                                   @Name(value = "limit", defaultValue = "0") Long limit) {
-        try (Coll coll = getMongoColl(hostOrKey, db, collection, compatibleValues)) {
-            return coll.find(query, project, sort, skip, limit).map(MapResult::new); // .onClose(coll::safeClose);
-        } catch (Exception e) {
-            log.error("apoc.mongodb.find - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], project = [" + project + "], sort = [" + sort + "], compatibleValues = [" + compatibleValues + "], skip = [" + skip + "], limit = [" + limit + "]",e);
-            throw new RuntimeException(e);
-        }
+        return executeMongoQuery(hostOrKey, db, collection, compatibleValues,
+                coll -> coll.find(query, project, sort, skip, limit).map(MapResult::new),
+                e -> log.error("apoc.mongodb.find - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], project = [" + project + "], sort = [" + sort + "], compatibleValues = [" + compatibleValues + "], skip = [" + skip + "], limit = [" + limit + "]",e));
     }
 
     @Procedure
@@ -127,23 +121,17 @@ public class MongoDB {
     @Procedure
     @Description("apoc.mongodb.delete(host-or-port,db-or-null,collection-or-null,list-of-maps) - delete the given documents from the mongodb collection and returns the number of affected documents")
     public Stream<LongResult> delete(@Name("host") String hostOrKey, @Name("db") String db, @Name("collection") String collection, @Name("query") Map<String, Object> query) {
-        try (Coll coll = getMongoColl(hostOrKey, db, collection, false)) {
-            return Stream.of(new LongResult(coll.delete(query))); // .onClose(coll::safeClose);
-        } catch (Exception e) {
-            log.error("apoc.mongodb.delete - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "]",e);
-            throw new RuntimeException(e);
-        }
+        return executeMongoQuery(hostOrKey, db, collection, false,
+                coll -> Stream.of(new LongResult(coll.delete(query))),
+                e -> log.error("apoc.mongodb.delete - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "]",e));
     }
 
     @Procedure
     @Description("apoc.mongodb.update(host-or-port,db-or-null,collection-or-null,list-of-maps) - updates the given documents from the mongodb collection and returns the number of affected documents")
     public Stream<LongResult> update(@Name("host") String hostOrKey, @Name("db") String db, @Name("collection") String collection, @Name("query") Map<String, Object> query, @Name("update") Map<String, Object> update) {
-        try (Coll coll = getMongoColl(hostOrKey, db, collection, false)) {
-            return Stream.of(new LongResult(coll.update(query, update))); // .onClose(coll::safeClose);
-        } catch (Exception e) {
-            log.error("apoc.mongodb.update - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], update = [" + update + "]",e);
-            throw new RuntimeException(e);
-        }
+        return executeMongoQuery(hostOrKey, db, collection, false,
+                coll -> Stream.of(new LongResult(coll.update(query, update))),
+                e -> log.error("apoc.mongodb.update - hostOrKey = [" + hostOrKey + "], db = [" + db + "], collection = [" + collection + "], query = [" + query + "], update = [" + update + "]",e));
     }
 
     private String getMongoDBUrl(String hostOrKey) {
@@ -202,6 +190,21 @@ public class MongoDB {
                     throw new RuntimeException("Could not create MongoDBClientWrapper instance", e);
                 }
             }
+        }
+    }
+
+    private <T> Stream<T> executeMongoQuery(String hostOrKey, String db, String collection, boolean compatibleValues,
+                                            Function<Coll, Stream<T>> execute, Consumer<Exception> onError) {
+        Coll coll = null;
+        try {
+            coll = getMongoColl(hostOrKey, db, collection, compatibleValues);
+            return execute.apply(coll).onClose(coll::safeClose);
+        } catch (Exception e) {
+            if (coll != null) {
+                coll.safeClose();
+            }
+            onError.accept(e);
+            throw new RuntimeException(e);
         }
     }
 }

--- a/src/test/java/apoc/mongodb/MongoDBTest.java
+++ b/src/test/java/apoc/mongodb/MongoDBTest.java
@@ -11,6 +11,7 @@ import org.bson.Document;
 import org.junit.*;
 import org.neo4j.graphdb.GraphDatabaseService;
 import org.neo4j.graphdb.QueryExecutionException;
+import org.neo4j.graphdb.Result;
 import org.neo4j.test.TestGraphDatabaseFactory;
 import org.testcontainers.containers.Container;
 import org.testcontainers.containers.GenericContainer;
@@ -22,6 +23,7 @@ import java.util.Date;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
+import java.util.stream.LongStream;
 import java.util.stream.Stream;
 
 import static apoc.util.MapUtil.map;
@@ -44,15 +46,17 @@ public class MongoDBTest {
     private static GraphDatabaseService db;
     private static MongoCollection<Document> collection;
 
-    private static Date currentTime = new Date();
+    private static final Date currentTime = new Date();
 
-    private static long longValue = 10_000L;
+    private static final long longValue = 10_000L;
 
     private static String HOST = null;
 
     private static Map<String, Object> params;
 
     private long numConnections = -1;
+
+    private static final long NUM_OF_RECORDS = 10_000L;
 
     @BeforeClass
     public static void setUp() throws Exception {
@@ -75,7 +79,11 @@ public class MongoDBTest {
         MongoDatabase database = mongoClient.getDatabase("test");
         collection = database.getCollection("test");
         collection.deleteMany(new Document());
-        collection.insertOne(new Document(map("name", "testDocument", "date", currentTime, "longValue", longValue)));
+        LongStream.range(0, NUM_OF_RECORDS)
+                .forEach(i -> collection
+                        .insertOne(new Document(map("name", "testDocument",
+                                "date", currentTime, "longValue", longValue))));
+//        collection.insertOne(new Document(map("name", "testDocument", "date", currentTime, "longValue", longValue)));
         db = new TestGraphDatabaseFactory()
                 .newImpermanentDatabaseBuilder()
                 .newGraphDatabase();
@@ -135,21 +143,14 @@ public class MongoDBTest {
 
     @Test
     public void testGet()  {
-        TestUtil.testCall(db, "CALL apoc.mongodb.get({host},{db},{collection},null)", params, r -> {
-            Map doc = (Map) r.get("value");
-            assertNotNull(doc.get("_id"));
-            assertEquals("testDocument", doc.get("name"));
-        });
+        TestUtil.testResult(db, "CALL apoc.mongodb.get({host},{db},{collection},null)", params,
+                res -> assertResult(res));
     }
 
     @Test
     public void testGetCompatible() throws Exception {
-        TestUtil.testCall(db, "CALL apoc.mongodb.get({host},{db},{collection},null,true)", params, r -> {
-            Map doc = (Map) r.get("value");
-            assertNotNull(doc.get("_id"));
-            assertEquals("testDocument", doc.get("name"));
-            assertEquals(currentTime.getTime(), doc.get("date"));
-        });
+        TestUtil.testResult(db, "CALL apoc.mongodb.get({host},{db},{collection},null,true)", params,
+                res -> assertResult(res, currentTime.getTime()));
     }
 
     @Test
@@ -163,33 +164,45 @@ public class MongoDBTest {
 
     @Test
     public void testFind() throws Exception {
-        TestUtil.testCall(db, "CALL apoc.mongodb.find({host},{db},{collection},{name:'testDocument'},null,null)", params, r -> {
+        TestUtil.testResult(db, "CALL apoc.mongodb.find({host},{db},{collection},{name:'testDocument'},null,null)",
+                params, res -> assertResult(res));
+    }
+
+    private void assertResult(Result res) {
+        assertResult(res, currentTime);
+    }
+
+    private void assertResult(Result res, Object date) {
+        int count = 0;
+        while (res.hasNext()) {
+            ++count;
+            Map<String, Object> r = res.next();
             Map doc = (Map) r.get("value");
             assertNotNull(doc.get("_id"));
             assertEquals("testDocument", doc.get("name"));
-        });
+            assertEquals(date, doc.get("date"));
+            assertEquals(longValue, doc.get("longValue"));
+        }
+        assertEquals(NUM_OF_RECORDS, count);
     }
 
     @Test
     public void testFindSort() throws Exception {
-        TestUtil.testCall(db, "CALL apoc.mongodb.find({host},{db},{collection},{name:'testDocument'},null,{name:1})", params, r -> {
-            Map doc = (Map) r.get("value");
-            assertNotNull(doc.get("_id"));
-            assertEquals("testDocument", doc.get("name"));
-        });
+        TestUtil.testResult(db, "CALL apoc.mongodb.find({host},{db},{collection},{name:'testDocument'},null,{name:1})",
+                params, res -> assertResult(res));
     }
 
     @Test
     public void testCount() throws Exception {
         TestUtil.testCall(db, "CALL apoc.mongodb.count({host},{db},{collection},{name:'testDocument'})", params, r -> {
-            assertEquals(1L, r.get("value"));
+            assertEquals(NUM_OF_RECORDS, r.get("value"));
         });
     }
 
     @Test
     public void testCountAll() throws Exception {
         TestUtil.testCall(db, "CALL apoc.mongodb.count({host},{db},{collection},null)", params, r -> {
-            assertEquals(1L, r.get("value"));
+            assertEquals(NUM_OF_RECORDS, r.get("value"));
         });
     }
 
@@ -197,7 +210,7 @@ public class MongoDBTest {
     public void testUpdate() throws Exception {
         TestUtil.testCall(db, "CALL apoc.mongodb.update({host},{db},{collection},{name:'testDocument'},{`$set`:{age:42}})", params, r -> {
             long affected = (long) r.get("value");
-            assertEquals(1L, affected);
+            assertEquals(NUM_OF_RECORDS, affected);
         });
     }
 

